### DISCUSSION TITLE
Is it possible to type output of child states?

### DISCUSSION THREAD

**Discussion by WIStudent** [statelyai/xstate#5015] [Q&A]
The documentation describes how to type the root output of a machine [here](https://stately.ai/docs/output#output-and-typescript), but it is also possible to set output on final states of a child state. Take a look at the following example:
```typescript
import { setup } from "xstate";

interface ChildOutput {
  result: "success" | "failure"
}

setup({}).createMachine({
  initial: "parentInitial",
  states: {
    parentInitial: {
      on: {
        NEXT: "child"
      }
    },
    child: {
      initial: "childInitial",
      states: {
        childInitial: {
          on: {
            SUCCESS: "childSuccess",
            FAILURE: "childFailure"
          }
        },
        childSuccess: {
          type: "final",
          output: (): ChildOutput => ({
            result: "success"
          })
        },
        childFailure: {
          type: "final",
          output: (): ChildOutput => ({
            result: "failure"
          })
        }
      },
      onDone: [
        {
          target: "parentA",
          guard: ({event}) => event.output.result === "success"  // Results in ts error because output is unknown
        },
        {
          target: "parentB",
          guard: ({event}) => (event.output as ChildOutput).result === "failure" // My current workaround
        },
        {
          target: "parentC"
        }
      ]
    },
    parentA: {},
    parentB: {},
    parentC: {}
  }
});
```
[Link to playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgZwKYwK5jgXzgMyghDgCIAPZGAQxlVIG4AoJ4AOzqn2oGNU4AwgAtgAGwAmAeQwwwMxEzhwoqZBlEwAXGTU8+yZKTgAfMtzEYVpJjhZpMYABQIcASgB0PFbVQBZXiJsqM6KcOzAMMDUotqkYNQqHACSbBFRoqQANKFUPsjaCKFK8YkwKWnRBUVKcBBsVTWNcAByAKIAGgAqsTwiEtZNuNU42Y29YuINTeGRlWTjEuWzGaNNuXT5CoNKC+JL6VPbtfVbRzUAygCqAgKt5+c9feLnGHqqhqtncABiAIJJABlLgAlVqPCbfagWKzVQa2I4jWE1XYvN4GQ7bGAATzAqFi+HY0SySMaEBkci0cEcrm0wgm0lk8gAvAA+KmFL5KFRqDSxXT6QwkmpuIWIo67SHQvGnI7Y3H4wkrIVKMmMynU2lPBkUuCs9nKmrc9SU0jmUSWegGkUIpFipp1AAidWlAG0hRyzjQoABzdCxEqoDi-Ylfb0YBKTdmoABugZgbl1bJjcfcqop7iNGl1TKZOleAqMcAA9EW4MDVMbkGE2HAYFXUFAiFA4AAjVA8agYNC1cnyYBVjBsADWbAgAHc2KLPk0PbKEr6TQGOAAhENnMMR7TOZMcBN6xw7mCp3vwahVukSbUyDyZ+A53OmqHmqzF0u+LFwHiWUpwMfQIcJGSbDiFO7oGl6C7+gkcYCAM2zwk0AC6wzTkuMC-AUdrFNBK6YahOEwAImGhLYbjMEAA)

The child state has 2 final states, `childSuccess` and `childFailure`, that set the `result` field in the output to `success` or `failure`. The parent uses guards in `onDone` to decide the next state based on the `result` field in the output. Unfortunately the type of the event that the guard receives is always `DoneStateEvent<unknown>`. Is there a way to configure the type of the child state output so that the event the guard receives is correctly typed? Or is my only option here to cast inside the guard?

**Comment by crishoj** [comment]
Also curious to know this.

Typing output at the topmost level is possible with `setup`:

```ts
setup({
  types: {
    output: {} as ChildOutput
  }
}).createMachine({
  // ...
})
```

But I am not aware of a way to specify types of "inline" child actors.
